### DISCUSSION TITLE
[mini.deps] support luarocks?

### DISCUSSION THREAD

**Discussion by jamestansx** [nvim-mini/mini.nvim#1387] [Q&A] [answered]
Hi!

Not sure if this has brought up, but would there be any plan to support luarocks?
Because currently the blocker of using `mini.deps` fully is [neorg](https://github.com/nvim-neorg/neorg) which required to install its dependencies using luarocks.

**Comment by echasnovski** [comment, marked as answer]
No, I am afraid luarocks support out of the box is not planned. Mostly for complexity reasons.

There are [various hooks](https://github.com/echasnovski/mini.nvim/blob/73bbcbfa7839c4b00a64965fb504f87461abefbd/doc/mini-deps.txt#L278-L289) for users to take advantage of. Actual way of utilizing that (how to make sure that proper rocks are installed) is also left to users. Maybe a custom `vim.system()` call will be enough?

**Reply by jamestansx** [reply]
> There are [various hooks](https://github.com/echasnovski/mini.nvim/blob/73bbcbfa7839c4b00a64965fb504f87461abefbd/doc/mini-deps.txt#L278-L289) for users to take advantage of. Actual way of utilizing that (how to make sure that proper rocks are installed) is also left to users. Maybe a custom vim.system() call will be enough?

This is my attempt, currently able to install `neorg`:

```lua
local add = MiniDeps.add

-- luarocks support
local luarocks = path_package .. "pack/deps/opt/luarocks/"
local lr_path  = luarocks .. "share/lua/5.1/"
local lr_cpath = luarocks .. "lib/lua/5.1/"
package.path = table.concat({
    package.path,
    lr_path .. "?.lua",
    lr_path .. "?/init.lua",
}, ";")
package.cpath = table.concat({
    package.cpath,
    lr_cpath .. "?.so",
}, ";")

local function install_luarocks(args)
    if vim.fn.executable("luarocks") ~= 1 then return end

    local rockspec = vim.fs.find(function(name, _)
        return name:match(".*%-1.rockspec$")
    end, { path = args.path, type = "file", limit = 1 })

    vim.system({
        "luarocks",
        "--tree", luarocks,
        "install",
        "--deps-only",
        "--lua-version", "5.1",
        "--deps-mode", "one",
        rockspec[1],
    }):wait()
end

local lr_add = function(tbl)
    tbl["hooks"] = {
        post_install = install_luarocks,
        post_checkout = install_luarocks,
    }

    add(tbl)
end

lr_add({
    source = 'nvim-neorg/neorg',
})
```

However, now when I do `DepsClean`, the `luarocks` directory would always appear in the list. Is there a way to always ignore a particular directory, or I would need to move the luarocks packages tree to other places?

**Reply by echasnovski** [reply]
> However, now when I do `DepsClean`, the `luarocks` directory would always appear in the list. Is there a way to always ignore a particular directory, or I would need to move the luarocks packages tree to other places?

I'd suggest using some other path, because both 'pack/deps/opt' and 'pack/deps/start' are special. My suggestion would be either 'pack/deps' directly, some other optional package directory ('pack/luarocks/opt', 'pack/luarocks/start', 'pack/luarocks'), or indeed another subdirectory of standard path. I'd prefer trying in that same order until it works (probably all should work).

**Reply by jamestansx** [reply]
Noted, thanks for the help!
